libs/utils/src/properties_encoding.c

```diff
@@ -0,0 +1,580 @@
+        return ENOMEM;
+    }
+
+    for (int i = 0; i < celix_arrayList_size(entry->typed.arrayValue); ++i) {
+        celix_array_list_entry_t arrayEntry = celix_arrayList_getEntry(entry->typed.arrayValue, i);
+        celix_array_list_element_type_t elType = celix_arrayList_getElementType(entry->typed.arrayValue);
+        json_t* jsonValue;
+        celix_status_t status = celix_properties_arrayElementEntryValueToJson(elType, arrayEntry, &jsonValue);
+        if (status != CELIX_SUCCESS) {
+            return status;
+        } else if (!jsonValue) {
+            // ignore unset values
+        } else {
+            int rc = json_array_append_new(array, jsonValue);
+            if (rc != 0) {
+                celix_err_push("Failed to append json string to array.");
+                return ENOMEM;
+            }
+        }
+    }
+
+    *out = celix_steal_ptr(array);
+    return CELIX_SUCCESS;
+}
+
+static celix_status_t
+celix_properties_entryValueToJson(const char* key, const celix_properties_entry_t* entry, int flags, json_t** out) {
+    *out = NULL;
+    switch (entry->valueType) {
+    case CELIX_PROPERTIES_VALUE_TYPE_STRING:
+        *out = json_string(entry->value);
+        break;
+    case CELIX_PROPERTIES_VALUE_TYPE_LONG:
+        *out = json_integer(entry->typed.longValue);
+        break;
+    case CELIX_PROPERTIES_VALUE_TYPE_DOUBLE:
+        if (isnan(entry->typed.doubleValue) || isinf(entry->typed.doubleValue)) {
+            if (flags & CELIX_PROPERTIES_ENCODE_ERROR_ON_NAN_INF) {
+                celix_err_pushf("Invalid NaN or Inf in key '%s'.", key);
+                return CELIX_ILLEGAL_ARGUMENT;
+            }
+            return CELIX_SUCCESS; // ignore NaN and Inf
+        }
+        *out = json_real(entry->typed.doubleValue);
+        break;
+    case CELIX_PROPERTIES_VALUE_TYPE_BOOL:
+        *out = json_boolean(entry->typed.boolValue);
+        break;
+    case CELIX_PROPERTIES_VALUE_TYPE_VERSION:
+        return celix_properties_versionToJson(entry->typed.versionValue, out);
+    case CELIX_PROPERTIES_VALUE_TYPE_ARRAY_LIST:
+        return celix_properties_arrayEntryValueToJson(key, entry, flags, out);
+    default:
+        // LCOV_EXCL_START
+        celix_err_pushf("Unexpected properties entry type %d.", entry->valueType);
+        return CELIX_ILLEGAL_ARGUMENT;
+        // LCOV_EXCL_STOP
+    }
+
+    if (!*out) {
+        celix_err_pushf("Failed to create json value for key '%s'.", key);
+        return ENOMEM;
+    }
+    return CELIX_SUCCESS;
+}
+
+static celix_status_t celix_properties_addJsonValueToJson(json_t* value, const char* key, json_t* obj, int flags) {
+    if (!value) {
+        // ignore unset values
+        return CELIX_SUCCESS;
+    }
+
+    json_t* field = json_object_get(obj, key);
+    if (field) {
+        if (flags & CELIX_PROPERTIES_ENCODE_ERROR_ON_COLLISIONS) {
+            celix_err_pushf("Invalid key collision. key '%s' already exists.", key);
```

Comment:
I could not trigger the error situation. I think it is reachable, but maybe not because the collision test cases always trigger in `celix_properties_addPropertiesEntryToJson`

Reply:
I added `TEST_F(PropertiesSerializationTestSuite, KeyCollision)` for this boundary condition.